hackforla/tdm-calculator
# Adds end-to-end feature for archiving and deleting user accounts

### **Summary** 
This PR solves #1395 
Implemented an end-to-end feature, creating a workflow that allows security admins to archive, un-archive/restore, and delete users along with their corresponding projects. 

### **Features** 
- **Archiving Users:** Security admins can now archive users. 
- **Un-archiving/Restoring Users:** Security admins can also un-archive and restore users.
- **Deleting Users and Projects:** Security admins can permanently delete users and the user's respective projects.
- **UI Enhancements:** Added buttons and confirmation modals to ensure admins understand the implications of their actions.

### **Technical Details** 
- Created a migration script to alter the tables "Login" and "Project" to add an `archivedAt` column that is set to `NULL` by default or to a timestamp if queried
- Created new, protected API endpoints: 
    - Accounts: 
        - PUT `/api/accounts/:id/archiveaccount`
        - PUT `/api/accounts/:id/unarchiveaccount`
        - GET `/api/accounts/archivedaccounts`
        - DELETE `/api/accounts/:id/deleteaccount`
    - Projects:
        -   GET `/api/projects/archivedprojects`
- Added new components in the frontend to handle the workflow.
- added new frontend routes to show all archived users or all archived projects

### **How to Test** 
1. Login as a security admin.
2. Navigate to the `/Roles` page.
3. Next to each user, there should be an 'options' button(vertical ellipsis).
4. Clicking the 'options' button will render a PopUp that can be clicked to archive the user. This will populate the `archivedAt` field for the user and their projects; thus, 'archiving' the user. Security Admins cannot be archived or deleted. The user is also not able to self-archive or self-delete.
5.  To see the archived users, navigate to the `/archivedaccounts` by following the link. Here, the security admin can 'un-archive and restore' a user OR permanently delete by clicking on the arrow or trash icons and their respective PopUp's. These actions will take effect on the user and their corresponding projects.
6. To see all archived projects, navigate to the `/archivedprojects` by following the link. This page only shows all archived projects - no actions. To return to the `/Roles` page, follow the link on the top of the page. 

### **Feature Notes**
- Deleting a user is irreversible - there is a currently a multi-step-process to delete a user to prevent accidental removal.
    1. select 'options'
    2. use popup to confirm user is to be archived
    3. Visit the archive page
    4. select 'delete'
    5. use popup to confirm user is to be permanently deleted
- Only Security Admins can archive, un-archive/restore, or delete users.
- The logged-in user cannot self-archive or self-delete
- Archived users cannot login

### **Future Improvements** 
- We could consider adding 2-Factor Auth such as email confirmation or PIN to delete users

### **Visuals** 
#### 


![firefox_uOiRVYKpGB](https://github.com/hackforla/tdm-calculator/assets/73868258/c4e29a23-fa1b-494e-95b3-82871ec2e405)
